src/components/submission/diff-item.vue

```diff
@@ -0,0 +1,226 @@
+          {{ $t(`editCaption.${typeOfChange}`) }}
+        </div>
+        <div class="nested-changes">
+          <submission-diff-item v-for="(change, index) in nestedDiffs" :key="index" :entry="change" :fields="fields" :parent-path="entry.path"/>
+        </div>
+      </template>
+    </div>
+    </div>
+  </div>
+</template>
+
+<script>
+export default {
+  name: 'SubmissionDiffItem',
+  props: {
+    entry: {
+      type: Object,
+      required: true
+    },
+    parentPath: {
+      type: Array,
+      default() {
+        return [];
+      }
+    },
+    fields: {
+      type: Object,
+      required: false
+    }
+  },
+  computed: {
+    isAtomicChange() {
+      // Is this a diff of a single field or whole subtree
+      if (this.entry.old === null)
+        if (typeof (this.entry.new) === 'object')
+          return false;
+      if (this.entry.new === null)
+        if (typeof (this.entry.old) === 'object')
+          return false;
+      return true;
+    },
+    visiblePath() {
+      if (!this.isAtomicChange)
+        return this.entry.path;
+      return this.entry.path.slice(0, this.entry.path.length - 1);
+    },
+    fieldName() {
+      return this.entry.path[this.entry.path.length - 1];
+    },
+    typeOfChange() {
+      if (this.entry.old === null)
+        return 'added';
+      if (this.entry.new === null)
+        return 'deleted';
+      return 'changed';
+    },
+    nestedDiffs() {
+      return this.flattenDiff(this.entry.new || this.entry.old, this.typeOfChange);
+    },
+    isBinary() {
+      // Compares path as array by converting it to /field1/field2 string and checks
+      // if it is for a binary field
+      const fullPath = this.parentPath.concat(this.entry.path);
+      const fullPathStr = fullPath.map((field) => (Array.isArray(field) ? field[0] : field)).join('/');
+      const basicPath = `/${fullPathStr}`;
+      if (this.fields[basicPath])
+        return true;
+      return false;
+    }
+  },
+  methods: {
+    flattenDiff(node, typeOfChange, prefix = []) {
+      // Flattens the inner diff (JSON of an entire added/removed tree structure)
+      // by recursively traversing the tree
+      let changes = [];
+      // eslint-disable-next-line guard-for-in
+      for (const k in node) {
+        const path = prefix.concat(k);
+        if (Array.isArray(node)) {
+          path.pop(); // pop off the index (e.g. 1 of toy #1)
+          path.pop(); // pop off the name (e.g. toy of toy #1)
+          path.push([prefix[prefix.length - 1], k]);
+        }
+        if (typeof (node[k]) === 'object') {
+          changes = changes.concat(this.flattenDiff(node[k], typeOfChange, path));
+        } else if (typeOfChange === 'added') {
+          changes.push({ path, old: null, new: node[k] });
+        } else {
+          changes.push({ path, old: node[k], new: null });
+        }
+      }
+      return changes;
+    },
+    binaryHref(value) {
+      return `/v1${this.$route.fullPath}/attachments/${value}`;
```

Comment:
You can access `projectId`, `xmlFormId`, and `instanceId` in order to use the `apiPaths` helper:

**Option 1:** `$this.route.params` (not preferred)

```js
this.$route.params.projectId
this.$route.params.xmlFormId
this.$route.params.instanceId
```

**Option 2:** the store

```js
computed: {
  ...requestData(['form', 'submission']),
  binaryHref(value) {
    return apiPaths.submissionAttachment(
      this.form.projectId,
      this.form.xmlFormId,
      false,
      this.submission.instanceId,
      value
    );
  }
}
```

Reply:
I'm trying this (**option 2** in particular) but `form` doesn't seem to be in the store! The `projectId` and `xmlFormId` are passed around as props to the main components, but they are not plumbed deep enough to get to the components I need them in. I'm temporarily falling back to **option 1** for the form fields.

Reply:
Ah you're totally right. `SubmissionShow` doesn't send a request for the form, because it generally doesn't need it, but that means that `form` won't exist. Other components on the page that use `apiPaths` do so by receiving `projectId` and `xmlFormId` as props. You could definitely consider doing so as well. But I do see that that'd require a lot of plumbing, passing those props from `SubmissionActivity` to `SubmissionFeedEntry` to `SubmissionDiffContainer` to `SubmissionDiffItem`. Maybe that's a reason to consider removing `SubmissionDiffContainer`, now that it's doing less?

I think it's also fine to continue using `this.$route.params`. In general, Vue Router [encourages](https://router.vuejs.org/guide/essentials/passing-props.html) the use of props over `$route`. However, I think it'd be reasonable to use `$route` in this case if that seems easier! (And you could use `$route` for `instanceId` if that seems nicer.) The main thing is that when testing `SubmissionDiffItem` in isolation, you'll need to mock `$route`.

Reply:
I'm going to try removing `SubmissionDiffContainer` and getting the props through to `SubmissionDiffItem`!